sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/pubsub/TestPubsub.java

```diff
@@ -172,6 +193,65 @@ public void publish(List<PubsubMessage> messages) throws IOException {
     pubsub.publish(eventsTopicPath, outgoingMessages);
   }
 
+  /** Pull up to 100 messages from {@link #subscriptionPath()}. */
+  public List<PubsubMessage> pull() throws IOException {
+    return pull(100);
+  }
+
+  /** Pull up to {@code maxBatchSize} messages from {@link #subscriptionPath()}. */
+  public List<PubsubMessage> pull(int maxBatchSize) throws IOException {
+    return pubsub.pull(0, subscriptionPath, 100, true).stream()
```

Comment:
For the `pull` methods, two things that make sense to me are to have no default and/or to match the defaults of a public Pubsub library. Since `PubsubClient` here is just something we built for Beam's purposes, I wouldn't try to match its API. TBH it seems like the native Pubsub API described at https://cloud.google.com/pubsub/docs/pull might be a better fit to avoid writing your own pulling loop. I assume `PubsubClient` adds conveniences so this may not be a straightforward decision. I'm OK either way, but something to consider.

Reply:
I don't see anything in the public pubsub API that saves me from writing my own pulling loop. The closest thing to what I've done here is the [synchronous pull example](https://cloud.google.com/pubsub/docs/pull#synchronous_pull), but that just does a single pull, no loop to wait for N messages. It also acknowledges the messages, which I wasn't doing.. I've added that now.

All I really want is a simple way to wait for the subscription to receive N messages, and then make an assertion about their content, so I can fluently test pipelines that write to pubsub. The `pull` methods could be a private implementation detail, would that be better? Actually the one that defaults to maxBatchSize = 100 could be removed entirely in that case.

I could also look at switching this from `PubsubClient` to the public library, it doesn't look like it would be too hard.

Reply:
My impression from (very briefly) peeking at their docs was that you could set up an asynchronous pull. This is very low stakes, while the value of your change is high. So I would not block a merge on this. Let's just move forwards.

Reply:
Ah I see. I dismissed the asynchronous option since it would require some sort of synchronization, but that may be simpler. I can look at doing that as a follow-up.